Sources/SourceControl/RepositoryManager.swift

```diff
@@ -125,58 +88,64 @@ public class RepositoryManager {
     /// - Parameters:
     ///   - repository: The repository to look up.
     ///   - skipUpdate: If a repository is available, skip updating it.
+    ///   - observabilityScope: The observability scope
+    ///   - delegateQueue: Dispatch queue for delegate events
+    ///   - callbackQueue: Dispatch queue for callbacks
     ///   - completion: The completion block that should be called after lookup finishes.
     public func lookup(
         repository: RepositorySpecifier,
-        skipUpdate: Bool = false,
-        on queue: DispatchQueue,
-        completion: @escaping LookupCompletion
+        skipUpdate: Bool,
+        observabilityScope: ObservabilityScope,
+        delegateQueue: DispatchQueue,
+        callbackQueue: DispatchQueue,
+        completion: @escaping (Result<RepositoryHandle, Error>) -> Void
     ) {
-        self.lookupQueue.addOperation {
-            // First look for a cached version of the handle.
-            // Also check state file / storage for resiliency.
-            let cachedHandle = self.storage.fileExists() ? self.repositoriesLock.withLock {
-                return self.repositories[repository.location.description]
-            } : nil
-
-            if let handle = cachedHandle {
-                let result = LookupResult(catching: {
-                    let start = DispatchTime.now()
-                    // Update the repository when it is being looked up.
-                    let repo = try handle.open()
+        // wrap the callback in the requested queue
+        let completion = { result in callbackQueue.async { completion(result) } }
 
-                    // Skip update if asked to.
+        self.lookupQueue.addOperation {
+            let relativePath = repository.storagePath()
+            let repositoryPath = self.path.appending(relativePath)
+            let handle = RepositoryManager.RepositoryHandle(manager: self, repository: repository, subpath: relativePath)
+
+            // errors when trying to check if a repository already exists are legitimate
+            // and recoverable, and as such can be ignored
+            if (try? self.provider.repositoryExists(at: repositoryPath)) ?? false {
+                return completion(.init(catching: {
+                    // skip update if not needed
                     if skipUpdate {
                         return handle
                     }
-
-                    queue.async {
-                        self.delegate?.handleWillUpdate(handle: handle)
+                    // Update the repository when it is being looked up.
+                    let start = DispatchTime.now()
+                    delegateQueue.async {
+                        self.delegate?.willUpdate(repository: handle.repository)
                     }
-
-                    try repo.fetch()
-
+                    let repository = try handle.open()
```

Comment:
If this throws an exception, could there be a case where the `willUpdate` is called but the `didUpdate` never gets called?  If possible it would be nice to always have those balanced, so that an IDE can know when the operation ended (even if that's with an error).  I suppose if the error has enough metadata then it could figure out to not expect `didUpdate` but it still feels easier on the caller if they are balanced.

Reply:
To be clear, I think this code already had problems where it could throw an error about a repository after calling `didFetch` etc, but it would be great to make this more robust if possible so a delegate doesn't have to work around it.

Reply:
yes, this maintained the existing flow where open/fetch could fail after the "willFetch" delegate will be called but before "didFetch" is called

Reply:
Seems reasonable.  I think I have a bug report on going in and fixing that, which I can do as a follow-on PR.

Reply:
happy to also fix in this PR if you want to share the bug #

Reply:
That would be fantastic — thanks!  rdar://69382774